### ISSUE TITLE
The return value for update() is wrong in WebIDL

### ISSUE THREAD

**Issue by philipwalton** [w3c/ServiceWorker#1472]
The [current WebIDL for the `ServiceWorkerRegistration`](https://w3c.github.io/ServiceWorker/#serviceworkerregistration-interface) is incorrect; it shows `update()` returning `Promise<void>`.

<img width="857" alt="Screen Shot 2019-09-16 at 4 18 56 PM" src="https://user-images.githubusercontent.com/326742/64940018-d9f9c280-d89d-11e9-98d7-37703a8a5b79.png">

I believe this is a mistake since implementations all return the registration object, and the [`update()` algorithm](https://w3c.github.io/ServiceWorker/#dom-serviceworkerregistration-update) says to return the registration as well.

**Comment by yoshisatoyanagisawa** [comment]
Let me check what is written in WebIDLs.
Chromium: 
> `[CallWith=ScriptState, RaisesException] Promise<ServiceWorkerRegistration> update();`

(https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/service_worker/service_worker_registration.idl;l=26;drc=047c7dc4ee1ce908d7fea38ca063fa2f80f92c77)

Gecko:
>  ` [Throws, NewObject]`
>  `Promise<undefined> update();`

(https://github.com/mozilla/gecko-dev/blob/fefa0c86cf24e6afa9f4dbb3823d4d16b2713dc7/dom/webidl/ServiceWorkerRegistration.webidl#L28)

WebKit:
>  `[NewObject] Promise<undefined> update();`

(https://github.com/WebKit/WebKit/blob/f8eb4cf035d529c6e55ebfd28c92fdc9468981d5/Source/WebCore/workers/service/ServiceWorkerRegistration.idl#L46)

As @saschanaz pointed out, Gecko and WebKit returns `Promise<undefined>`.
However, although the WebIDL looks invalid, Chromium seems to follow the step by step description in [Resolve Job Promise algorithm](https://w3c.github.io/ServiceWorker/#resolve-job-promise-algorithm).

If mismatch of WebIDL and [Resolve Job Promise algorithm](https://w3c.github.io/ServiceWorker/#resolve-job-promise-algorithm) matters, adjusting the WebIDL can be a choice.  What do you think?

**Comment by saschanaz** [comment]
Depends on whether there's enough implementation interest for that, otherwise I'd vote for following the majority. cc @asutherland

**Comment by annevk** [comment]
Implementations not actually enforcing validity of this IDL construct through asserts or some such seems like a problem as well here. Separate, but still.

**Comment by annevk** [comment]
Fixed by #1767.

**Comment by preitinger** [comment]
> The [current WebIDL for the `ServiceWorkerRegistration`](https://w3c.github.io/ServiceWorker/#serviceworkerregistration-interface) is incorrect; it shows `update()` returning `Promise<void>`.
> 
> <img alt="Screen Shot 2019-09-16 at 4 18 56 PM" width="857" src="https://user-images.githubusercontent.com/326742/64940018-d9f9c280-d89d-11e9-98d7-37703a8a5b79.png">
> 
> I believe this is a mistake since implementations all return the registration object, and the [`update()` algorithm](https://w3c.github.io/ServiceWorker/#dom-serviceworkerregistration-update) says to return the registration as well.

No, I think the `update algorithm` does not say that it returns the job. It just says what is done with the job and then that the `promise` is returned. Nothing is said what is in this promise which means for me it is `void`. So it was correct imo!

**Comment by annevk** [comment]
It's not a matter of opinion.

You have to actually go through all the Job indirection (which could do with some refactoring) to see what the promise gets resolved with.

**Comment by saschanaz** [comment]
>It's not a matter of opinion.

Agreed, and just to clarify in case, what matters is web compatibility here, so we followed what implementations already do.

**Comment by yoshisatoyanagisawa** [comment]
I am not sure either [PR](https://github.com/w3c/ServiceWorker/pull/1771) or here to reply, but let me reply here.

As I mentioned in https://github.com/w3c/ServiceWorker/issues/1472#issuecomment-2840469985, update() returns the promise that will be resolved with `ServiceWorkerRegistration`.  Let me explain one by one.

This is the update() algorithm you might have already seen.
https://w3c.github.io/ServiceWorker/#service-worker-registration-update

In the algorithm, it creates `promise`, and the `promise` is passed to `CreateJob`.  The returned `job` has the promise inside.
https://w3c.github.io/ServiceWorker/#create-job

Then the `job` will be passed to ScheduleJob
https://w3c.github.io/ServiceWorker/#schedule-job

RunJob will eventually be called.
https://w3c.github.io/ServiceWorker/#run-job-algorithm

Since this is Update, Update will eventually be called.
https://w3c.github.io/ServiceWorker/#update

Install will eventually be called if not promise rejected.
https://w3c.github.io/ServiceWorker/#install

In Step 7, [Resolve Job Promise](https://w3c.github.io/ServiceWorker/#resolve-job-promise) is called.
In the algorithm, `job`'s promise will eventually be resolved by convertedValue, it is always be the return value of [the service worker registration object](https://w3c.github.io/ServiceWorker/#get-the-service-worker-registration-object), and you can see [ServiceWorkerRegistration](https://w3c.github.io/ServiceWorker/#serviceworkerregistration) is returned.
Therefore, it is always be `ServiceWorkerRegistration` or rejected.

Please feel free to ask questions if something not clear.

**Comment by annevk** [comment]
I do agree with Peter's comment in the PR that resolving with the this object is dumb, but it also seems far too involved to change at this point. We should just not do it for new methods.

**Comment by preitinger** [comment]
OK, this goes to far for me...
From my prospective as a user, I would never use any return value from this
function `update()` as it is obvious that even those who write the spec do
not know exactly what and why sth is to return. I know the typescript
signature definition for this function is not really relevant officially,
but for me as user, it will be relevant and I just will never use the
return value of this `update()` function and you can specify what ever is
correct. ;-)
Thank you

Am Mi., 21. Mai 2025 um 08:29 Uhr schrieb Anne van Kesteren <
***@***.***>:

**Comment by annevk** [comment]
The specification (unfortunately) requires that the this object is returned through the path @yoshisatoyanagisawa explained above and (unfortunately) that is what all browsers implement. IDL used to assert that nothing is returned (browsers unfortunately don't assert such a thing which is why this wasn't discovered until now). So the IDL did have a bug.
